File: geometry/python/vectors_test.py
import unittest
from vectors import Vec2, Vec3
from approximate import EPSILON


class TestMatrices(unittest.TestCase):
    def test_eq(self):
        self.assertTrue(Vec2(3, 4) == Vec2(3, 4))

    def test_approx_eq(self):
        ep = EPSILON
        # Vec2
        result = Vec2(3, 4).approx_eq(Vec2(3 + ep, 4 + ep))
        self.assertTrue(result)
        # Vec3
        result = Vec3(3, 4, 5).approx_eq(Vec3(3 + ep, 4, 5 + ep))
        self.assertTrue(result)

    def test_add(self):
        # Vec2
        expect = Vec2(4, 6)
        result = Vec2(1, 2) + Vec2(3, 4)
        self.assertEqual(expect, result)
        # Vec3
        expect = Vec3(4, 6, 2)
        result = Vec3(1, 2, 1) + Vec3(3, 4, 1)
        self.assertEqual(expect, result)

    def test_sub(self):
        # Vec2
        expect = Vec2(2, 2)
        result = Vec2(3, 4) - Vec2(1, 2)
        self.assertEqual(expect, result)
        # Vec3
        expect = Vec3(2, 2, 2)
        result = Vec3(3, 4, 5) - Vec3(1, 2, 3)
        self.assertEqual(expect, result)

    def test_negate(self):
        # Vec2
        expect = Vec2(-3, -4)
        result = -Vec2(3, 4)
        self.assertEqual(expect, result)
        # Vec3
        expect = Vec3(-3, -4, -5)
        result = -Vec3(3, 4, 5)
        self.assertEqual(expect, result)

    def test_invert(self):
        # Vec2
        expect = Vec2(0.5, 0.5)
        result = Vec2(2, 2).invert()
        self.assertEqual(expect, result)
        # Vec3
        expect = Vec3(0.5, 0.5, 0.5)
        result = Vec3(2, 2, 2).invert()
        self.assertEqual(expect, result)

    def test_magnitude(self):
        # Vec2
        expect = 5
        result = Vec2(3, 4).mag()
        self.assertEqual(expect, result)
        # Vec3
        expect = 7.0710678118654755
        result = Vec3(3, 4, 5).mag()
        self.assertEqual(expect, result)

    def test_scale(self):
        # Vec2
        expect = Vec2(33, 44)
        result = Vec2(3, 4).scale(11)
        self.assertEqual(expect, result)
        # Vec3
        expect = Vec3(33, 44, 55)
        result = Vec3(3, 4, 5).scale(11)
        self.assertEqual(expect, result)

    def test_dot_product(self):
        # Vec2
        expect = 11
        result = Vec2(3, 4).dot(Vec2(1, 2))
        self.assertEqual(expect, result)
        # Vec3
        expect = 26
        result = Vec3(3, 4, 5).dot(Vec3(1, 2, 3))
        self.assertEqual(expect, result)

    def test_distance(self):
        # Vec2
        expect = 7
        result = Vec2(8, 0).distance(Vec2(1, 0))
        self.assertEqual(expect, result)
        # Vec3
        expect = 7
        result = Vec3(10, 0, 0).distance(Vec3(3, 0, 0))

    def test_normalize(self):
        # Vec2
        expect = Vec2(0.6, 0.8)
        result = Vec2(3, 4).normalize()
        self.assertEqual(expect, result)
        # Vec3
        expect = Vec3(0.5883484054145521, 0.7844645405527362,
                      0.19611613513818404)
        result = Vec3(3, 4, 1).normalize()
        self.assertEqual(expect, result)

    def test_round(self):
        # Vec2
        expect = Vec2(1, 4)
        result = Vec2(0.75, 4.25).round()
        self.assertEqual(expect, result)
        # Vec3
        expect = Vec3(1, 4, 1)
        result = Vec3(0.75, 4.25, 1.15).round()
        self.assertEqual(expect, result)


if __name__ == "__main__":
    unittest.main()

File: geometry/golang/vectors/vectors_test.go
package vectors

import (
	"didact/geometry/golang/matrices"
	utils "didact/geometry/golang/utilities"
	"testing"
)

func TestApproxEq(t *testing.T) {
	tolerance := utils.Tolerance

	v1 := Vec2{3 + tolerance, 4 + tolerance}
	v2 := Vec2{3, 4}
	if v1.AboutEqual(v2) == false {
		t.Errorf("Test ApproxEq Vec2 failed: false negative. Check tolerance in test.")
	}
	// Push values past threshold for approximate equality
	// by pushing epsilon into one figure higher in significance.
	v1 = Vec2{3 + tolerance*10, 4 + tolerance}
	if v1.AboutEqual(v2) == true {
		t.Errorf("Test ApproxEq Vec2 failed: false positive. Check tolerance in test.")
	}
}

func TestAbsolute(t *testing.T) {
	expect := Vec2{3, 4}
	result := Vec2{-3, -4}.Absolute()
	if expect != result {
		t.Errorf("Test Absolute Vec2 failed. Expected: %v, Got: %v", expect, result)
	}
}

func TestAdd(t *testing.T) {
	expect := Vec2{2, 4}
	result := Vec2{4, 6}.Add(Vec2{-2, -2})
	if expect != result {
		t.Errorf("Test Add Vec2 failed. Expected: %v, Got: %v", expect, result)
	}
}

func TestSubtract(t *testing.T) {
	expect := Vec2{2, 2}
	result := Vec2{3, 4}.Subtract(Vec2{1, 2})
	if expect != result {
		t.Errorf("Test Subtract Vec2 failed. Expected: %v, Got: %v", expect, result)
	}
}

func TestNegate(t *testing.T) {
	expect := Vec2{-3, -4}
	result := Vec2{3, 4}.Negate()
	if expect != result {
		t.Errorf("Test Negate Vec2 failed. Expected: %v, Got: %v", expect, result)
	}
}

func TestInvert(t *testing.T) {
	expect := Vec2{0.5, 0.5}
	result := Vec2{2, 2}.Invert()
	if result != expect {
		t.Errorf("Test Invert Vec2 failed. Expected: %v, Got: %v", expect, result)
	}
}

func TestMagnitude(t *testing.T) {
	expect := 5.0
	result := Vec2{3.0, 4.0}.Magnitude()
	if expect != result {
		t.Errorf("Test Magnitude Vec2 failed. Expected: %.2f, Got: %.2f", expect, result)
	}
}

func TestScale(t *testing.T) {
	expect := Vec2{1, 0.5}
	result := Vec2{0.5, 0.25}.Scale(2.0)
	if expect != result {
		t.Errorf("Test Scale Vec2 failed. Expected: %v, Got: %v", expect, result)
	}
}

func TestDotProduct(t *testing.T) {
	expect := 11.0
	result := Vec2{3, 4}.DotProduct(Vec2{1, 2})
	if expect != result {
		t.Errorf("Test Dot Product Vec2 failed. Expected: %.2f, Got: %.2f", expect, result)
	}
}

func TestDistance(t *testing.T) {
	expect := 7.0
	result := Vec2{8, 0}.Distance(Vec2{1, 0})
	if expect != result {
		t.Errorf("Test Distance Vec2 failed. Expected: %v, Got: %v", expect, result)
	}
}

func TestInterpolate(t *testing.T) {
	expect := Vec2{2, 1.5}
	result := Vec2{1, 1}.Interpolate(0.5, Vec2{3, 2})
	if expect != result {
		t.Errorf("Test Interpolate Vec2 failed. Expected: %v, Got: %v", expect, result)
	}
}

func TestNormalize(t *testing.T) {
	expect := Vec2{0.6, 0.8}
	result := Vec2{3, 4}.Normalize()
	if expect != result {
		t.Errorf("Test Normalize Vec2 failed. Expected: %v, Got: %v", expect, result)
	}
}

func TestRound(t *testing.T) {
	expect := Vec2{1, 4}
	result := Vec2{0.75, 4.25}.Round()
	if expect != result {
		t.Errorf("Test Round Vec2 failed. Expected: %v, Got: %v", expect, result)
	}
}

func TestTransformBy(t *testing.T) {
	mat := matrices.Scale(10, 10)
	expect := Vec2{30, 40}
	result := Vec2{3, 4}.TransformBy(mat)
	if expect != result {
		t.Errorf("Test Vec2 TransformBy failed. Expected: %v, Got: %v", expect, result)
	}
}

func TestStringer(t *testing.T) {
	expect := "Vec2(3, 4)"
	result := Vec2{3.0, 4.0}.String()
	if expect != result {
		t.Errorf("Test Stringer Vec2 failed. Expected: %v, Got: %v", expect, result)
	}
}
